### PR TITLE
Critical LB dump fixes

### PR DESCRIPTION
**Pull request by mayhem**
When the listenbrainz-cron-prod container is killed while dumps are being sync'ed to the FTP server, empty dump files may be left in the container which may end up being sync'ed to the FTP server. This has caused 0B *private* dump files to appear on the FTP server. While no private data has been leaked as part of this, we need to ensure that the dump creation process is more robust than it is now. This PR aims to accomplish this.

### CODE REVIEW

File: listenbrainz/db/dump_manager.py [hunk 1]
```diff
@@ -157,12 +168,24 @@ def create_incremental(location, threads, dump_id):
...
+        except IOError as e:
+            sys.exit(-1)
 
         # if in production, send an email to interested people for observability
         send_dump_creation_notification(dump_name, 'incremental')
 
+        # Write the DUMP_ID file so that the FTP sync scripts can be more robust
+        print(os.path.join(dump_path, "DUMP_ID.txt"))
+        with open(os.path.join(dump_path, "DUMP_ID.txt"), "w") as f:
+            f.write("%s %s incremental\n" % (end_time.strftime('%Y%m%d-%H%M%S'), dump_id))
```
Comment (zas) [inline listenbrainz/db/dump_manager.py:185]: Why a newline here and not in full? https://github.com/metabrainz/listenbrainz-server/pull/1178/files#diff-e4f7644db199bd1cedfa7296a724b02679b34ea30b8920fdbae2eff15153cd9cR129

**Review by zas — CHANGES_REQUESTED** [review]

File: listenbrainz/db/dump_manager.py (hunk 1 above)
Comment (mayhem) [reply listenbrainz/db/dump_manager.py:185]: Good catch. fixing.

Comment (mayhem) [reply listenbrainz/db/dump_manager.py:185]: Fixed

File: listenbrainz/db/dump_manager.py [hunk 2]
```diff
@@ -157,12 +168,24 @@ def create_incremental(location, threads, dump_id):
...
+            if not sanity_check_dumps(dump_path, 6):
+                return sys.exit(-1)
+        except IOError as e:
+            sys.exit(-1)
 
         # if in production, send an email to interested people for observability
         send_dump_creation_notification(dump_name, 'incremental')
 
+        # Write the DUMP_ID file so that the FTP sync scripts can be more robust
+        print(os.path.join(dump_path, "DUMP_ID.txt"))
```
Comment (zas) [inline listenbrainz/db/dump_manager.py:183]: Is this `print()` statement useful?

File: listenbrainz/db/dump_manager.py [hunk 3]
```diff
@@ -280,6 +306,37 @@ def write_hashes(location):
...
+        expected_count (int): the number of files that are expected to be present
+    Return:
+        boolean: true if the dump passes the sanity check
+    """
+
+    count = 0
+    for file in os.listdir(location):
+        try:
+            dump_file = os.path.join(location, file)
+            print(dump_file)
```
Comment (zas) [inline listenbrainz/db/dump_manager.py:323]: No use of logger here?

File: listenbrainz/db/dump_manager.py [hunk 4]
```diff
@@ -280,6 +306,37 @@ def write_hashes(location):
...
+            if os.path.getsize(dump_file) == 0:
+                print("Dump file %s is empty!" % dump_file)
+                return False
+            count += 1
+        except IOError as e:
```
Comment (zas) [inline listenbrainz/db/dump_manager.py:328]: `os.path.getsize()` may raise `OSError`.

In Python3, `os` module always raise `OSError` (but IOError is a deprecated alias), see https://www.python.org/dev/peps/pep-3151/#confusing-set-of-os-related-exceptions about this subject.

https://docs.python.org/3/library/os.path.html#os.path.getsize

File: listenbrainz/db/dump_manager.py [hunk 5]
```diff
@@ -280,6 +306,37 @@ def write_hashes(location):
             raise
 
 
+def sanity_check_dumps(location, expected_count):
+    """ Sanity check the generated dummps to ensure that none are empty
```
Comment (zas) [inline listenbrainz/db/dump_manager.py:310]:
```suggestion
    """ Sanity check the generated dumps to ensure that none are empty and the expected count was created.
```

File: listenbrainz/db/dump_manager.py [hunk 6]
```diff
@@ -157,12 +168,24 @@ def create_incremental(location, threads, dump_id):
             write_hashes(dump_path)
         except IOError as e:
             current_app.logger.error('Unable to create hash files! Error: %s', str(e), exc_info=True)
-            return
+            sys.exit(-1)
+
+        try:
+            if not sanity_check_dumps(dump_path, 6):
+                return sys.exit(-1)
+        except IOError as e:
```
Comment (zas) [inline listenbrainz/db/dump_manager.py:176]: See my comment about `OSError` vs `IOError` below.

File: listenbrainz/db/dump_manager.py [hunk 7]
```diff
@@ -111,14 +111,25 @@ def create_full(location, threads, dump_id, last_dump_id):
...
+            sys.exit(-1)
 
         # if in production, send an email to interested people for observability
         send_dump_creation_notification(dump_name, 'fullexport')
 
         current_app.logger.info('Dumps created and hashes written at %s' % dump_path)
 
+        # Write the DUMP_ID file so that the FTP sync scripts can be more robust
+        with open(os.path.join(dump_path, "DUMP_ID.txt"), "w") as f:
+            f.write("%s %s full\n" % (end_time.strftime('%Y%m%d-%H%M%S'), dump_id))
```
Comment (zas) [inline listenbrainz/db/dump_manager.py:129]: `end_time.strftime('%Y%m%d-%H%M%S')` is done 3 times, perhaps store in a variable to reduce code redundancy.

**Review by zas — CHANGES_REQUESTED** [review]

File: listenbrainz/db/dump_manager.py (hunk 3 above)
Comment (mayhem) [reply listenbrainz/db/dump_manager.py:323]: The loggers are not available in this context, so I would have to create another logger for some of the code, some of the time. Not what I really want to be doing right now.

File: listenbrainz/db/dump_manager.py [hunk 8]
```diff
@@ -157,12 +169,23 @@ def create_incremental(location, threads, dump_id):
             write_hashes(dump_path)
         except IOError as e:
             current_app.logger.error('Unable to create hash files! Error: %s', str(e), exc_info=True)
-            return
+            sys.exit(-1)
+
+        try:
+            if not sanity_check_dumps(dump_path, 6):
+                return sys.exit(-1)
+        except IOError as e:
```
Comment (zas) [inline listenbrainz/db/dump_manager.py:177]: `os.listdir()` raises `OSError` (in fact, all `os.` methods do), think about changing those, those are aliases now, but they may not in the future.

File: listenbrainz/db/dump_manager.py [hunk 9]
```diff
@@ -275,11 +301,39 @@ def write_hashes(location):
...
                 sha256sum = subprocess.check_output(['sha256sum', os.path.join(location, file)]).decode('utf-8').split()[0]
                 f.write(sha256sum)
-        except IOError as e:
+        except OSError as e:
             current_app.logger.error('IOError while trying to write hash files for file %s: %s', file, str(e), exc_info=True)
             raise
 
 
+def sanity_check_dumps(location, expected_count):
+    """ Sanity check the generated dumps to ensure that none are empty
```
Comment (zas) [inline listenbrainz/db/dump_manager.py:310]: You discarded part of my proposed change, but the method doesn't only check if none of the files are empty, but also check if the expected count matches.

**Review by zas — APPROVED** [review]

**Pull request merged**